cpp/src/arrow/compute/kernels/scalar_nested.cc

```diff
@@ -429,6 +429,169 @@ const FunctionDoc make_struct_doc{"Wrap Arrays into a StructArray",
                                   {"*args"},
                                   "MakeStructOptions"};
 
+struct MapArrayLookupFunctor {
+  static Result<int64_t> FindOneMapValueIndex(const Array& keys, const Scalar& query_key,
+                                              const int64_t start, const int64_t end,
+                                              const bool from_back = false) {
+    if (!from_back) {
+      for (int64_t idx = start; idx < end; ++idx) {
+        ARROW_ASSIGN_OR_RAISE(std::shared_ptr<Scalar> key, keys.GetScalar(idx));
+
+        if (key->Equals(query_key)) return idx;
+      }
+    } else {
+      for (int64_t idx = end - 1; idx >= start; --idx) {
+        ARROW_ASSIGN_OR_RAISE(std::shared_ptr<Scalar> key, keys.GetScalar(idx));
+
+        if (key->Equals(query_key)) return idx;
+      }
+    }
+    return -1;
+  }
+
+  static Result<std::shared_ptr<Scalar>> GetScalarOutput(KernelContext* ctx,
+                                                         const MapScalar map_scalar) {
```

Comment:
Is this even remotely close to what you had in mind @lidavidm ?
Also, is there a rough example I could follow for `VisitArrayValuesInline` and `UnboxScalar` you mentioned for templating the kernel?

Reply:
- So, this was the basic structure I was thinking of,
```cpp
  template <typename KeyType>
  struct MapArrayLookupFunctor {
  .
  .
  .
  const KeyType query_key = UnboxScalar<KeyType>::Unbox(*options.query_key);
  ...
  const std::shared_ptr<Array> keys = struct_array.field(0);
  ...
  
  VisitArrayValuesInline<KeyType>(
    *keys->data(),
    [&](KeyType key) -> Status {
      if (key == query_key) {
        if (occurrence == FIRST) {
          ...
          return Status::Cancelled("Found key");
        } else /*occurrence == ALL*/ { 
            ...
            return Status::OK(); 
        }
      } else { return Status::OK(); }
     },
    [&]() -> Status {
      ...
      return Status::OK();
    }
  );
  ``` 
- But as the values are visited _inline_, how would we deal with the `LAST` option (iterating from the back)?
- And just so I'm clear, when you say templating the kernel, you mean templating for the `KeyType`, right?

Reply:
For LAST, you could keep track of the index outside the "loop" and simply overwrite it each time (though that is perhaps not as efficient), returning the final index. We could later define a visitor that goes in reverse.

Reply:
And yes, templating is for the type of the key.